Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ExamGroupsController.cs

```diff
@@ -0,0 +1,286 @@
+    using Resource = Resources.Areas.Administration.ExamGroups.ExamGroupsController;
+    using ViewModelType = OJS.Web.Areas.Administration.ViewModels.ExamGroups.ExamGroupAdministrationViewModel;
+
+    public class ExamGroupsController : LecturerBaseGridController
+    {
+        private const int DefaultUsersTakeCount = 20;
+        private const int DefaultContestsToTake = 15;
+
+        private readonly IExamGroupsDataService examGroupsData;
+        private readonly IUsersDataService usersData;
+        private readonly IContestsDataService contestsData;
+
+        public ExamGroupsController(
+            IOjsData data,
+            IExamGroupsDataService examGroupsData,
+            IUsersDataService usersData,
+            IContestsDataService contestsData)
+            : base(data)
+        {
+            this.examGroupsData = examGroupsData;
+            this.usersData = usersData;
+            this.contestsData = contestsData;
+        }
+
+        public ActionResult Index() => this.View();
+
+        public override IEnumerable GetData()
+        {
+            var examGroups = this.examGroupsData.GetAll();
+
+            if (this.User.IsLecturer())
+            {
+                examGroups = examGroups.Where(eg => eg.Contest == null ||
+                    (eg.Contest.Lecturers.Any(l => l.LecturerId == this.UserProfile.Id) ||
+                        eg.Contest.Category.Lecturers.Any(l => l.LecturerId == this.UserProfile.Id)));
+            }
+
+            return examGroups.Select(ViewModelType.FromExamGroup);
+        }
+
+        public override object GetById(object id) => this.GetByIdAsNoTracking((int)id);
+
+        [HttpPost]
+        public ActionResult Create([DataSourceRequest]DataSourceRequest request, ViewModelType model)
+        {
+            var contestId = model.ContestId.HasValue
+                ? this.contestsData.GetByIdQuery(model.ContestId.Value).Select(c => c.Id).FirstOrDefault()
+                : default(int);
+
+            if (contestId != default(int))
+            {
+                if (!this.UserHasContestRights(contestId))
+                {
+                    this.ModelState.AddModelError(nameof(model.ContestId), Resource.Cannot_attach_contest);
+                    return this.GridOperation(request, model);
+                }
+
+                this.BaseCreate(model.GetEntityModel());
+            }
+            else
+            {
+                this.ModelState.AddModelError(nameof(model.ContestId), string.Empty);
+            }
+            
+            return this.GridOperation(request, model);
+        }
+
+        [HttpPost]
+        public ActionResult Update([DataSourceRequest]DataSourceRequest request, ViewModelType model)
+        {
+            if (!model.Id.HasValue)
+            {
+                return this.GridOperation(request, model);
+            }
+
+            var entity = this.GetByIdAsNoTracking(model.Id.Value);
```

Comment:
Could we have better naming here? `model`  is somewhat fine but `entity` starts muddying things up